### PR TITLE
Add/Edit Product: Fix transient product name while adding images.

### LINKED ISSUE #5739: Add/Edit Product: product title changes are reset after adding a new product image
**Describe the bug**

When creating or editing a product, product title changes are reset after adding a product image for the first time (either taken from the camera or selected from the device library).

**To Reproduce**

- Go to the products tab
- Tap on an editable product or tap "+" in the navigation bar to add a product of any type
- Edit the product title
- Tap on the "+" cell or any image cell in the images header to edit product images
- Tap "Add Photos"
- Tap "Choose from device" or "Take a photo" to upload a photo --> after going back to the product form, notice that the product title changes are reset

**Screenshots**

https://user-images.githubusercontent.com/1945542/147209976-ebcaaac6-a307-4e6d-9bd9-41569f77c1bb.mov

**Expected behavior**
The product title changes should not be reset after adding a new image.

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: any
- iOS version: any
- WooCommerce iOS version: 8.2

**WordPress Environment**
<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin.
```
</details>

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #5739

### Description
**Issue**

In both Add and Edit Product screens, if images are added to a product after editing the product name (before saving), the `tableView` shows the old product name for a short duration. We can also observe that the edited product name reappears in a while after the images are loaded.

**Why?**
- The `product` (from `viewModel`) will change if the user types and changes the name. But, we are NOT recreating `tableViewModel` and reloading the `tableView` in this case. You can check comments in `func observeProductName` to know why. [More info in PR](https://github.com/woocommerce/woocommerce-ios/pull/3832)
- The above scenario results in `tableViewModel` and `product` getting out of sync. i.e. `product` has name changed in it, but `tableViewModel` doesn’t reflect the "changed name".
- When `onImageStatusesUpdated` calls `reconfigureDataSource` without recreating `tableViewModel` we end up showing old `product` information in the `tableView`.
- Note that the name information isn’t lost. It lives inside `product`. When we eventually recreate `tableViewModel` and reload using `reconfigureDataSource` we have the new product name. (This is probably why we haven’t noticed this bug before.)

**Solution**
- By recreating `tableViewModel` using the current `product` before calling `reconfigureDataSource`, we are making sure that we are not showing outdated `product` information in `tableView`.

### Testing instructions

**Add product flow**
1. Navigate to the Products tab.
1. Tap on + button on the top right.
1. Choose a product type from the action sheet and navigate to Add product screen.
1. Type a title in the Add product screen.
1. Tap on "Add a product image". 
1. Tap "Add Photos".
1. Tap "Choose from device" or "Take a photo" to upload a photo.
1. After going back to the "Add product screen" observe that you can see the edited name.

**Edit product flow**
1. Ensure that you have an editable product available.
1. Navigate to the Products tab.
1. Tap on an editable product and edit the name.
1. Tap on the "+" cell or any image cell in the images header to edit product images
1. Tap "Add Photos".
1. Tap "Choose from device" or "Take a photo" to upload a photo.
1. After going back to the "Add product screen" observe that you can see the edited name.

### Screen recordings
| Flow | Before | After |
| --- | --- | --- |
| Add product | [Add-Before](https://user-images.githubusercontent.com/524475/148665547-59dd4107-7e26-4de5-afbb-2c16ba2e6cd9.mov) | [Add-After](https://user-images.githubusercontent.com/524475/148665533-5c61b3cb-ddae-4188-a61f-f4233a66584c.mov) |
| Edit product | [Edit-Before](https://user-images.githubusercontent.com/524475/148665566-c93c9e66-e5a2-4125-9b71-b7d76db40a9c.mov) |  [Edit-After](https://user-images.githubusercontent.com/524475/148665574-589ce5f7-33cf-4903-b721-7cdf883e59fc.mov) |

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by itsmeichigo — COMMENTED** [review]
Thank you for the detailed explanation of the issue. And your solution fixes the problem at hand. 

However, I'm concerned about how the `observeProductName()` method works. I mean I'm not sure if there's really a case when the product name is changed outside of the screen. Do you think it's possible to update the method instead - to maybe detect when a user finishes editing the product name to update the product accordingly?

Also - I think we should update the release notes for this fix too.

**Comment by selanthiraiyan** [comment]
> I mean I'm not sure if there's really a case when the product name is changed outside of the screen. Do you think it's possible to update the method instead - to maybe detect when a user finishes editing the product name to update the product accordingly?


@Ecarrion Could you kindly help us understand when the product name will be changed outside of the `ProductFormViewController`? Thank you!

@itsmeichigo  I am thinking about doing the following,

1. Replace `var productName: Observable<String>?` in `ProductFormViewModelProtocol` with a Combine Publisher
1.  Use a `debounce` operator to wait for the user to finish typing
1. Call `func onProductUpdated(product: ProductModel)` to reload the tableView. 

What do you think?

**Comment by itsmeichigo** [comment]
@selanthiraiyan great idea asking @Ecarrion to clarify his intention in the original method 👍 

As for your proposed solution, I'm thinking using `debounce` and then reloading the table view may cause another issue when, say, the user stops typing for a second, they may lose their focus from the text field. I haven't tried this though so I'm not 100% sure.

Is there any way to observe the text field and get noticed when it is out of focus? That way we can safely reload the table view - WDYT?

Then again, if it's too complicated, we can keep your solution to avoid over-engineering. We'll have to wait for Ernesto's explanation too, he might have better insight as to why `observeProductName` was implemented the way it is.

**Comment by selanthiraiyan** [comment]
> Is there any way to observe the text field and get noticed when it is out of focus? That way we can safely reload the table view - WDYT?

It should be possible by adding a listener to the `ProductFormTableViewDataSource` and IMO this looks like a great solution for this problem. 

> Then again, if it's too complicated, we can keep your solution to avoid over-engineering. We'll have to wait for Ernesto's explanation too, he might have better insight as to why observeProductName was implemented the way it is.

I don't think that your solution will get too complicated to implement. But, yes, let us wait for Ernesto's inputs, to understand more about the other scenario of "product name" being changed from outside the screen.

**Comment by Ecarrion** [comment]
Hi @itsmeichigo and @selanthiraiyan, that `observeProductName()` method exists to cover the following case.

- User changes the product name locally
- User creates an attribute or a variation (This updates the whole product, overriding the unsaved product name)
- ViewModel detects that there was a pending name change and fires an event to the name observable
- View re-renders re un-saved product name and updates the save button state.

Does that cover your inquiries?

**Comment by itsmeichigo** [comment]
@Ecarrion Thanks for the clarification! So the change that happens outside of the screen is actually made programmatically when the user creates an attribute or a variation (they're on the Variations screen at that moment). I think it might be worth it to update the comment of `observeProductName()` of this specific case to make it clearer to understand 😅

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift [hunk 1]
```diff
@@ -528,6 +528,9 @@ private extension ProductFormViewController {
     }
 
     func onImageStatusesUpdated(statuses: [ProductImageStatus]) {
+        tableViewModel = DefaultProductFormTableViewModel(product: product,
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:559]: Due to the complicated cases of product editing, I think it's best to add clear comments on why we added this line. Maybe we should explain here that we need to update the view model to make sure the displayed product is update-to-date at this point?

**Review by itsmeichigo — COMMENTED** [review]
@selanthiraiyan thanks to Ernesto's explanation, I think it's best to keep the implementation as-is for now. Could you also update the comment on `observeProductName()` with an extra line explaining that the method helps with the case when a user adds or edit variations?

I also left a comment below asking for more documentation on the code. I think we can benefit from clear comments to avoid accidentally removing something important 😅

**Comment by selanthiraiyan** [comment]
@Ecarrion Thanks a lot for the details! 🙇 

@itsmeichigo I have added comments to both methods. Kindly take a look when you can. And, I have updated the release notes now. 🙏

**Review by itsmeichigo — APPROVED** [review]
Thank you @selanthiraiyan for adding really detailed comments!

**Pull request merged**